### PR TITLE
Extract ThemingScope interop and move it to Common

### PR DESCRIPTION
**Pull request by hughbe**
## Proposed Changes
- Extract ThemingScope interop 
- Move ThemingScope to its own file

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2331)

### CODE REVIEW

File: src/Common/src/ThemingScope.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        ///  Activate() does nothing if a theming context is already active on the current thread, which is good
+        ///  for perf reasons. However, in some cases, like in the Timer callback, we need to put another context
+        ///  on the stack even if one is already present. In such cases, this method helps - you get to manage
+        ///  the cookie yourself though.
+        /// </summary>
+        public static IntPtr Activate()
+        {
+            IntPtr userCookie = IntPtr.Zero;
+
+            if (Application.UseVisualStyles && s_contextCreationSucceeded && OSFeature.Feature.IsPresent(OSFeature.Themes))
```
Comment (RussKie) [inline src/Common/src/ThemingScope.cs]: Here and below please invert ifs to reduce nesting

Comment (hughbe) [reply src/Common/src/ThemingScope.cs]: Done

File: src/Common/src/ThemingScope.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            return false;
+        }
+
+        /// <summary>
+        ///  Activate() does nothing if a theming context is already active on the current thread, which is good
+        ///  for perf reasons. However, in some cases, like in the Timer callback, we need to put another context
+        ///  on the stack even if one is already present. In such cases, this method helps - you get to manage
+        ///  the cookie yourself though.
+        /// </summary>
+        public static IntPtr Activate()
```
Comment (JuditRose) [inline src/Common/src/ThemingScope.cs]: What do you think about this? (I cannot add a multiline suggestion)
```suggestion
        public static IntPtr Activate()
            {
                IntPtr userCookie = IntPtr.Zero;
                if (IsContextCreatedButNotActivated())
                {
                    userCookie = ActivateActCtx(hActCtx, out userCookie) ? userCookie : IntPtr.Zero;
                }
                return userCookie;
            }

            private static bool IsContextCreatedButNotActivated()
            {
                return Application.UseVisualStyles && contextCreationSucceeded && OSFeature.Feature.IsPresent(OSFeature.Themes) && !IsContextActive();
            }
```

File: src/Common/src/ThemingScope.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        private static IntPtr s_hActCtx;
+        private static bool s_contextCreationSucceeded;
+
+        /// <summary>
+        ///  We now use explicitactivate everywhere and use this method to determine if we
+        ///  really need to activate the activationcontext.  This should be pretty fast.
+        /// </summary>
+        private static bool IsContextActive()
+        {
+            if (s_contextCreationSucceeded && Kernel32.GetCurrentActCtx(out IntPtr current).IsTrue())
```
Comment (JuditRose) [inline src/Common/src/ThemingScope.cs]:
```suggestion
            return s_contextCreationSucceeded && Kernel32.GetCurrentActCtx(out IntPtr current).IsTrue() && current == s_hActCtx;
```

File: src/Common/src/ThemingScope.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+            return userCookie;
+        }
+
+        /// <summary>
+        ///  Use this to deactivate a context activated by calling ExplicitActivate.
+        /// </summary>
+        public static IntPtr Deactivate(IntPtr userCookie)
+        {
+            if (userCookie == IntPtr.Zero || Kernel32.DeactivateActCtx(0, userCookie).IsTrue())
```
Comment (JuditRose) [inline src/Common/src/ThemingScope.cs]: Same here as above I think it would be nice to decouple the condition from the action

Comment (hughbe) [reply src/Common/src/ThemingScope.cs]: Feel like its a bit of an overkill here but elsewhere I agree

**Pull request merged**